store/src/migration.rs

```diff
@@ -0,0 +1,212 @@
+
+#[derive(Debug)]
+struct Migration {
+    /// The name of the migration.
+    name: &'static str,
+    /// The version this migration was released (anything below this should be migrated).
+    _version: &'static str,
+}
+
+/// The list of migrations that need to be run in order to get the database up to date and start the server.
+/// Add new migrations to the end of the list.
+fn data_migrations() -> &'static Vec<Migration> {
+    MIGRATIONS.get_or_init(Vec::new)
+}
+
+#[derive(Debug, Clone, sqlx::FromRow)]
+// We want to retrieve these fields for logging but we don't refer to them directly
+#[allow(dead_code)]
+struct Version {
+    id: i64,
+    version: String,
+    installed_at: chrono::NaiveDateTime,
+}
+
+/// The data migrator is responsible for running data migrations the node requires as part of the verison upgrade.
+pub struct DataMigrator {
+    prev_version: Option<Version>,
+    required_migrations: Vec<&'static Migration>,
+    pool: SqlitePool,
+}
+
+impl DataMigrator {
+    /// Create a new data migrator. This updates the version table with the current version and determine which migrations need to be run.
+    pub async fn try_new(pool: SqlitePool) -> Result<Self> {
+        let current_version = env!("CARGO_PKG_VERSION").to_string();
+
+        let prev_version: Option<Version> = sqlx::query_as(
+            "SELECT id, version, installed_at FROM ceramic_one_version ORDER BY installed_at DESC limit 1;",
+        )
+        .fetch_optional(pool.reader())
+        .await?;
+
+        sqlx::query(
+            "INSERT INTO ceramic_one_version (version) VALUES ($1) ON CONFLICT DO NOTHING;",
+        )
+        .bind(&current_version)
+        .execute(pool.writer())
+        .await?;
+
+        let applied_migrations = DataMigration::fetch_all(&pool).await?;
+        debug!(?prev_version, %current_version, ?applied_migrations, "Current data migration status");
+        // In the future, we can filter out migrations that are not required based on the version as well
+        let required_migrations = data_migrations()
+            .iter()
+            .flat_map(|candidate| {
+                if applied_migrations
+                    .iter()
+                    .find(|m| m.name == candidate.name)
+                    .and_then(|m| m.completed_at)
+                    .is_some()
+                {
+                    None
+                } else {
+                    Some(candidate)
+                }
+            })
+            .collect();
+
+        tracing::debug!("required migrations: {:?}", required_migrations);
+
+        Ok(Self {
+            prev_version,
+            required_migrations,
+            pool,
+        })
+    }
+
+    /// Determines whether migrations are needed. Will mark all migrations as "complete" on a fresh install.
+    pub async fn needs_migration(&self) -> Result<bool> {
+        let new_install = self.is_new_install().await?;
+
+        if new_install {
```

Comment:
nitpick: this feels like a weird place for this logic to live.  Seems like there should be an `init` method or something like that that is called once at startup that handles this, rather than bundling logic into this boolean accessor method.

That said this migration code is going to be fairly self-contained and mostly only run during startup anyway, so it probably doesn't matter that much.

Reply:
Yeah, agreed. I changed it to happen at start up and refactored some things into functions.. sorry, I started with a one line change and then couldn't help myself.

Thinking about this made me realize that this should all be behind &mut and that ceramic one should probably expose a `CeramicDataMigration` service or something so the binary is completely unaware of all store crate things. I started making those changes and backed them out because that's something for another PR.